packages/xo-server-perf-alert/src/RrdHostVm.js

```diff
@@ -0,0 +1,357 @@
+          value = Math.round((100 * (total - free)) / total)
+          break
+        }
+        default:
+          throw new Error(`Can't compute alert of type ${rule.variableName} for VM`)
+      }
+
+      if (rule.isTriggeredBy(value)) {
+        vmAlarm.push(new Alarm({ rule, target: xoVm, value }))
+      }
+    }
+    return vmAlarm
+  }
+  /**
+   * @param {XoHost} xoHost
+   * @param {Object} hostStat
+   *
+   * @returns {Array<Alarm>}
+   */
+  #computeAlarms(xoHost, hostStat) {
+    /**
+     * @type {Array<Alarm>}
+     */
+
+    let alarms = this.#computeHostAlarm(xoHost, hostStat, this.#rules.getObjectAlerts(xoHost))
+    for (const [vmUid, vmStats] of Object.entries(hostStat.vms ?? {})) {
+      const xoVm = this.#xo.getObject(vmUid)
+      alarms = alarms.concat(this.#computeVmAlarm(xoVm, vmStats, this.#rules.getObjectAlerts(xoVm)))
+    }
+    return alarms
+  }
+
+  /**
+   *
+   * @returns {Promise<Array<Alarm>>}
+   */
+
+  async computeActiveAlarms() {
+    // make queries host by host, that way we don't need to cache RRD
+    /**
+     * @type {Array<XoHost>}
+     */
+
+    const hosts = Object.values(this.#xo.objects.indexes.type.host ?? {})
+    /**
+     * @type {Map<XoHost['id'], Set<XoVm['id']>}
+     */
+    const watchedVmsByHosts = new Map()
+    Object.values(this.#xo.objects.indexes.type.VM ?? {})
+
+      .filter(vm => this.#rules.isObjectAffected(vm))
+      .forEach(vm => {
+        if (!watchedVmsByHosts.has(vm.$container)) {
+          watchedVmsByHosts.set(vm.$container, new Set([vm.id]))
+        } else {
+          watchedVmsByHosts.get(vm.$container).add(vm.id)
+        }
+      })
+
+    const alarmsByObjects = []
+
+    await asyncEach(
+      hosts,
+      async host => {
+        const withHostData = this.#rules.isObjectAffected(host)
+        let vmUuid
+
+        const vms = watchedVmsByHosts.get(host.id)
+
+        if (vms === undefined) {
+          vmUuid = 'none'
+        } else {
+          if (vms.size === 1) {
+            ;[vmUuid] = vms
+          } else {
+            vmUuid = 'all'
+          }
+        }
+        if (withHostData || vmUuid !== 'none') {
+          const hostStats = await this.#getHostStats(host, this.#rules.getMaxPeriod(), { withHostData, vmUuid })
+          alarmsByObjects.push(this.#computeAlarms(host, hostStats))
+        }
+      },
+      { concurrency: 5 }
+    )
+
+    return [].concat(...alarmsByObjects)
+  }
+
+  async #poll(onChanges, delay) {
+    try {
+      // ensure we at least wait for the delay
+      // but handle gracefully long running computation
+      if (this.#lastChangeComputation) {
+        const nextRun = this.#lastChangeComputation + delay - Date.now()
+        await new Promise(resolve => {
+          const interval = setTimeout(() => {
+            resolve()
+          }, nextRun)
+          this.#abortWaitController.signal.addEventListener('abort', () => {
+            clearInterval(interval)
+          })
```

Comment:
Hm, is that possible to have an abort event before the promise resolve?
If yes, is that mean the promise will stay pending?

If we receive an abort signal, we clean the `setTimeout`, so I suppose the `resolve()` fn will never be triggered, right?